### PR TITLE
feat: Add enterprise billing usage endpoints and response types

### PR DESCRIPTION
**Pull request by maishivamhoo123**
## Summary
Adds support for the GitHub Enterprise billing usage API endpoints.
#4262 

## New methods on BillingService
GetEnterpriseUsageReport
GetEnterpriseUsageSummary
GetEnterprisePremiumRequestUsageReport
GetEnterpriseAICreditUsage
 
And added test for all the new End points .

### CODE REVIEW

File: github/enterprise_billing.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+type EnterpriseUsageReport struct {
+	UsageItems []*EnterpriseUsageItem `json:"usageItems,omitempty"`
+}
+
+// GetEnterpriseUsageReport returns a report of the total usage for an enterprise using the enhanced billing platform.
+//
+// GitHub API docs: https://docs.github.com/enterprise-cloud@latest/rest/billing/usage?apiVersion=2022-11-28#get-billing-usage-report-for-an-enterprise
+//
+//meta:operation GET /enterprises/{enterprise}/settings/billing/usage
+func (s *BillingService) GetEnterpriseUsageReport(ctx context.Context, enterprise string, opts *EnterpriseUsageReportOptions) (*EnterpriseUsageReport, *Response, error) {
```
Comment (alexandear) [inline github/enterprise_billing.go]:
```suggestion
func (s *EnterpriseService) GetUsageReport(ctx context.Context, enterprise string, opts *EnterpriseUsageReportOptions) (*EnterpriseUsageReport, *Response, error) {
```

**Comment by maishivamhoo123** [comment]
@alexandear  @gmlewis  @Not-Dhananjay-Mishra   i made all the changes can you please review it ?
Thank you.

File: github/enterprise_billing.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+	Day          int    `url:"day,omitempty"`
+	Organization string `url:"organization,omitempty"`
+	User         string `url:"user,omitempty"`
+	Model        string `url:"model,omitempty"`
+	Product      string `url:"product,omitempty"`
+	CostCenterID string `url:"cost_center_id,omitempty"`
+}
+
+// EnterpriseUsageSummaryOptions specifies optional parameters for the BillingService.GetEnterpriseUsageSummary method.
+type EnterpriseUsageSummaryOptions struct {
```
Comment (alexandear) [inline github/enterprise_billing.go:31]: All these option structs have similar fields:
Image

I think we can embed `EnterpriseUsageReportOptions` into `EnterprisePremiumRequestUsageReportOptions` and `EnterpriseUsageSummaryOptions`.

File: github/enterprise_billing_test.go [hunk 3]
```diff
@@ -0,0 +1,1018 @@
...
+	}
+	if report.UsageItems[0].Product != "Actions" {
+		t.Errorf("Expected first product to be Actions, got %v", report.UsageItems[0].Product)
+	}
+	if report.UsageItems[1].Product != "Copilot" {
+		t.Errorf("Expected second product to be Copilot, got %v", report.UsageItems[1].Product)
+	}
+}
+
+func TestEnterpriseService_GetUsageSummary_WithRepository(t *testing.T) {
```
Comment (alexandear) [inline github/enterprise_billing_test.go]: Why do we need this test? It duplicates `TestEnterpriseService_GetUsageSummary_WithAllFilters`

File: github/enterprise_billing_test.go
Comment (alexandear) [inline github/enterprise_billing_test.go:1]: Please review all tests and try not adding redundant tests that check the same functionality.

File: github/enterprise_billing.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+	return report, resp, nil
+}
+
+// GetUsageSummary returns a summary report of usage for an enterprise.
+//
+// GitHub API docs: https://docs.github.com/enterprise-cloud@latest/rest/billing/usage?apiVersion=2022-11-28#get-billing-usage-summary-for-an-enterprise
+//
+//meta:operation GET /enterprises/{enterprise}/settings/billing/usage/summary
+func (s *EnterpriseService) GetUsageSummary(ctx context.Context, enterprise string, opts *EnterpriseUsageSummaryOptions) (*EnterpriseAggregatedUsageReport, *Response, error) {
```
Comment (Not-Dhananjay-Mishra) [inline github/enterprise_billing.go]: Response schema of `GetUsageSummary` is little bit different from `GetPremiumRequestUsageReport` and `GetAICreditUsage`. Its response schema doesn't contain `user` and `model`, In my opinion we should make a new struct for it.

**Comment by maishivamhoo123** [comment]
@gmlewis  @alexandear  @Not-Dhananjay-Mishra  All tests and lint checks are passing in my local setup.
Screenshot 2026-06-11 153028

However, I'm not sure why the CI is failing here. Could you please guide me on how to debug this issue and identify the root cause? I'd also appreciate any suggestions on how to fix the linting error.
Thank you!

**Comment by alexandear** [comment]
> However, I'm not sure why the CI is failing here. Could you please guide me on how to debug this issue and identify the root cause? I'd also appreciate any suggestions on how to fix the linting error. Thank you!

Don't worry; it's a GitHub problem. @gmlewis will retry the workflow run.

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @maishivamhoo123!

The new endpoints are missing tests that make calls to `testBadOptions` and `testNewRequestAndDoFailure`. For example:

```
const methodName = "ListSelectedReposForOrgSecret"
testBadOptions(t, methodName, func() (err error) {
        _, _, err = client.Dependabot.ListSelectedReposForOrgSecret(ctx, "\n", "\n", opts)
        return err
})

testNewRequestAndDoFailure(t, methodName, client, func() (*Response, error) {
        got, resp, err := client.Dependabot.ListSelectedReposForOrgSecret(ctx, "o", "NAME", opts)
        if got != nil {
                t.Errorf("testNewRequestAndDoFailure %v = %#v, want nil", methodName, got)
        }
        return resp, err

        })
})
```

There are many examples in this repo. Please add similar tests to fix code coverage.

File: github/enterprise_billing.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+	Year  int  `json:"year"`
+	Month *int `json:"month,omitempty"`
+	Day   *int `json:"day,omitempty"`
+}
+
+// EnterpriseAggregatedUsageItem represents a single usage line item in an enterprise billing premium request or AI credit report.
+type EnterpriseAggregatedUsageItem struct {
+	Product          string  `json:"product"`
+	SKU              string  `json:"sku"`
+	Model            *string `json:"model,omitempty"`
```
Comment (alexandear) [inline github/enterprise_billing.go]: [`ghec.yaml`](https://github.com/github/rest-api-description/blob/main/descriptions/ghec/ghec.yaml) lists `model` in the `required` set for both billing-premium-request-usage-report-ghe and billing-ai-credit-usage-report-ghe item schemas

```suggestion
	Model            string `json:"model"`
```

**Comment by maishivamhoo123** [comment]
@alexandear  and @gmlewis  i made all the recommended changes can you please review this .
Thank you!

**Review by gmlewis — APPROVED** [review]
Thank you, @maishivamhoo123!
LGTM.
Awaiting LGTM+Approval from other reviewers before merging:

cc: @alexandear - @Not-Dhananjay-Mishra

File: github/enterprise_billing_test.go [hunk 6]
```diff
@@ -0,0 +1,1026 @@
...
+
+	ctx, cancel := context.WithCancel(t.Context())
+	cancel()
+
+	_, resp, err := client.Enterprise.GetUsageReport(ctx, "test-enterprise", nil)
+	if err == nil {
+		t.Error("GetUsageReport should return error with canceled context")
+	}
+	_ = resp
+}
```
Comment (alexandear) [inline github/enterprise_billing_test.go]: Why do we need this test function? There are no occurrences of `context.WithCancel` in other tests.

Comment (alexandear) [reply github/enterprise_billing_test.go]: I propose removing `TestEnterpriseService_*_CanceledContext` functions for consistency.

File: github/enterprise_billing_test.go [hunk 7]
```diff
@@ -0,0 +1,1026 @@
...
+	testNewRequestAndDoFailure(t, methodName, client, func() (*Response, error) {
+		got, resp, err := client.Enterprise.GetUsageReport(ctx, "test-enterprise", opts)
+		if got != nil {
+			t.Errorf("testNewRequestAndDoFailure %v = %#v, want nil", methodName, got)
+		}
+		return resp, err
+	})
+}
+
+func TestEnterpriseService_GetUsageReport_WithCostCenter(t *testing.T) {
```
Comment (alexandear) [inline github/enterprise_billing_test.go]: I propose to include `CostCenterID: "cc-123"` into the `opts` in the `TestEnterpriseService_GetUsageReport` and remove this test.

File: github/enterprise_billing_test.go [hunk 8]
```diff
@@ -0,0 +1,1026 @@
...
+func TestEnterpriseService_GetUsageReport_InvalidEnterprise(t *testing.T) {
+	t.Parallel()
+	client, _, _ := setup(t)
+
+	ctx := t.Context()
+	_, _, err := client.Enterprise.GetUsageReport(ctx, "%", nil)
+	testURLParseError(t, err)
+}
+
+func TestEnterpriseService_GetUsageReport_NoOptions(t *testing.T) {
```
Comment (alexandear) [inline github/enterprise_billing_test.go]: This test is not needed and can be removed.

File: github/enterprise_billing_test.go [hunk 9]
```diff
@@ -0,0 +1,1026 @@
...
+	testNewRequestAndDoFailure(t, methodName, client, func() (*Response, error) {
+		got, resp, err := client.Enterprise.GetUsageSummary(ctx, "test-enterprise", opts)
+		if got != nil {
+			t.Errorf("testNewRequestAndDoFailure %v = %#v, want nil", methodName, got)
+		}
+		return resp, err
+	})
+}
+
+func TestEnterpriseService_GetUsageSummary_MultipleItems(t *testing.T) {
```
Comment (alexandear) [inline github/enterprise_billing_test.go]: This test is not needed and can be removed.

File: github/enterprise_billing_test.go [hunk 10]
```diff
@@ -0,0 +1,1026 @@
...
+	}
+	if report.UsageItems[0].Product != "Actions" {
+		t.Errorf("Expected first product to be Actions, got %v", report.UsageItems[0].Product)
+	}
+	if report.UsageItems[1].Product != "Copilot" {
+		t.Errorf("Expected second product to be Copilot, got %v", report.UsageItems[1].Product)
+	}
+}
+
+func TestEnterpriseService_GetUsageSummary_WithAllFilters(t *testing.T) {
```
Comment (alexandear) [inline github/enterprise_billing_test.go]: This test is not needed and can be removed.

File: github/enterprise_billing_test.go [hunk 11]
```diff
@@ -0,0 +1,1026 @@
...
+	_, resp, err := client.Enterprise.GetUsageSummary(ctx, "test-enterprise", nil)
+	if err == nil {
+		t.Error("GetUsageSummary should return error on 404 response")
+	}
+	if resp == nil {
+		t.Error("GetUsageSummary should return response even on error")
+	}
+}
+
+func TestEnterpriseService_GetUsageSummary_CanceledContext(t *testing.T) {
```
Comment (alexandear) [inline github/enterprise_billing_test.go]: This test is not needed and can be removed.

File: github/enterprise_billing_test.go [hunk 12]
```diff
@@ -0,0 +1,1026 @@
...
+
+	if report.Enterprise != "GitHub" {
+		t.Errorf("Expected enterprise GitHub, got %v", report.Enterprise)
+	}
+	if report.User == nil || *report.User != "testuser" {
+		t.Errorf("Expected user testuser, got %v", report.User)
+	}
+	if len(report.UsageItems) != 1 {
+		t.Errorf("Expected 1 usage item, got %v", len(report.UsageItems))
+	}
```
Comment (alexandear) [inline github/enterprise_billing_test.go]: Please use `cmp.Equal` instead.

File: github/enterprise_billing_test.go [hunk 13]
```diff
@@ -0,0 +1,1026 @@
...
+	testNewRequestAndDoFailure(t, methodName, client, func() (*Response, error) {
+		got, resp, err := client.Enterprise.GetPremiumRequestUsageReport(ctx, "test-enterprise", opts)
+		if got != nil {
+			t.Errorf("testNewRequestAndDoFailure %v = %#v, want nil", methodName, got)
+		}
+		return resp, err
+	})
+}
+
+func TestEnterpriseService_GetPremiumRequestUsageReport_WithAllOptions(t *testing.T) {
```
Comment (alexandear) [inline github/enterprise_billing_test.go]: This test is not needed and can be removed.

File: github/enterprise_billing_test.go [hunk 14]
```diff
@@ -0,0 +1,1026 @@
...
+	_, resp, err := client.Enterprise.GetPremiumRequestUsageReport(ctx, "test-enterprise", nil)
+	if err == nil {
+		t.Error("GetPremiumRequestUsageReport should return error on 401 response")
+	}
+	if resp == nil {
+		t.Error("GetPremiumRequestUsageReport should return response even on error")
+	}
+}
+
+func TestEnterpriseService_GetPremiumRequestUsageReport_CanceledContext(t *testing.T) {
```
Comment (alexandear) [inline github/enterprise_billing_test.go]: This test is not needed and can be removed.

File: github/enterprise_billing_test.go [hunk 15]
```diff
@@ -0,0 +1,1026 @@
...
+	}
+	if report.User == nil || *report.User != "testuser" {
+		t.Errorf("Expected user testuser, got %v", report.User)
+	}
+	if len(report.UsageItems) != 1 {
+		t.Errorf("Expected 1 usage item, got %v", len(report.UsageItems))
+	}
+	if report.UsageItems[0].UnitType != "credits" {
+		t.Errorf("Expected unitType credits, got %v", report.UsageItems[0].UnitType)
+	}
```
Comment (alexandear) [inline github/enterprise_billing_test.go]: Please use `cmp.Equal` instead.

File: github/enterprise_billing_test.go [hunk 16]
```diff
@@ -0,0 +1,1026 @@
...
+	testNewRequestAndDoFailure(t, methodName, client, func() (*Response, error) {
+		got, resp, err := client.Enterprise.GetAICreditUsage(ctx, "test-enterprise", opts)
+		if got != nil {
+			t.Errorf("testNewRequestAndDoFailure %v = %#v, want nil", methodName, got)
+		}
+		return resp, err
+	})
+}
+
+func TestEnterpriseService_GetAICreditUsage_FloatQuantities(t *testing.T) {
```
Comment (alexandear) [inline github/enterprise_billing_test.go]: This test is not needed and can be removed.

File: github/enterprise_billing_test.go [hunk 17]
```diff
@@ -0,0 +1,1026 @@
...
+
+	if report.UsageItems[0].GrossQuantity != 1500.5 {
+		t.Errorf("Expected GrossQuantity 1500.5, got %v", report.UsageItems[0].GrossQuantity)
+	}
+	if report.UsageItems[0].NetAmount != 14.0 {
+		t.Errorf("Expected NetAmount 14.0, got %v", report.UsageItems[0].NetAmount)
+	}
+}
+
+func TestEnterpriseService_GetAICreditUsage_WithCostCenter(t *testing.T) {
```
Comment (alexandear) [inline github/enterprise_billing_test.go]: This test is not needed and can be removed.

File: github/enterprise_billing_test.go [hunk 18]
```diff
@@ -0,0 +1,1026 @@
...
+	}
+	if report.CostCenter.ID != "cc-engineering" {
+		t.Errorf("Expected CostCenter ID cc-engineering, got %v", report.CostCenter.ID)
+	}
+	if report.CostCenter.Name != "Engineering Team" {
+		t.Errorf("Expected CostCenter Name Engineering Team, got %v", report.CostCenter.Name)
+	}
+}
+
+func TestEnterpriseService_GetAICreditUsage_WithOrganization(t *testing.T) {
```
Comment (alexandear) [inline github/enterprise_billing_test.go]: This test is not needed and can be removed.

File: github/enterprise_billing_test.go [hunk 19]
```diff
@@ -0,0 +1,1026 @@
...
+	_, resp, err := client.Enterprise.GetAICreditUsage(ctx, "test-enterprise", nil)
+	if err == nil {
+		t.Error("GetAICreditUsage should return error on 400 response")
+	}
+	if resp == nil {
+		t.Error("GetAICreditUsage should return response even on error")
+	}
+}
+
+func TestEnterpriseService_GetAICreditUsage_NoItems(t *testing.T) {
```
Comment (alexandear) [inline github/enterprise_billing_test.go]: This test is not needed and can be removed.

File: github/enterprise_billing_test.go [hunk 20]
```diff
@@ -0,0 +1,1026 @@
...
+	if resp == nil {
+		t.Error("GetAICreditUsage returned nil response")
+	}
+
+	if len(report.UsageItems) != 0 {
+		t.Errorf("Expected 0 usage items, got %v", len(report.UsageItems))
+	}
+}
+
+func TestEnterpriseService_GetAICreditUsage_CanceledContext(t *testing.T) {
```
Comment (alexandear) [inline github/enterprise_billing_test.go]: This test is not needed and can be removed.

File: github/enterprise_billing_test.go [hunk 21]
```diff
@@ -0,0 +1,1026 @@
...
+	cancel()
+
+	_, resp, err := client.Enterprise.GetAICreditUsage(ctx, "test-enterprise", nil)
+	if err == nil {
+		t.Error("GetAICreditUsage should return error with canceled context")
+	}
+	_ = resp
+}
+
+func TestEnterpriseService_GetAICreditUsage_WithAllOptions(t *testing.T) {
```
Comment (alexandear) [inline github/enterprise_billing_test.go]: This test is not needed and can be removed.

**Comment by maishivamhoo123** [comment]
@alexandear Thank you for your review .
I made all the suggested changes can you please take a another look .
Thank you !

**Review by Not-Dhananjay-Mishra — APPROVED** [review]
LGTM

**Review by alexandear — APPROVED** [review]

**Comment by gmlewis** [comment]
Thank you, @maishivamhoo123, @alexandear, and @Not-Dhananjay-Mishra!
Merging.
